File: src/systems/LevelGenerator.ts

import { Planet, PlanetType } from '../entities/Planet';
import { FuelItem } from '../entities/FuelItem';
import { GameConfig } from '../engine/GameConfig';

export class LevelGenerator {
    private visitedChunks: Set<string> = new Set();
    private lastFuelSpawnY: number = 0;
    private spawnInterval: number = GameConfig.fuelSpawnInterval;

    generateItems(shipY: number): FuelItem[] {
        const items: FuelItem[] = [];

        // shipY is negative as we go up.
        if (shipY < this.lastFuelSpawnY - this.spawnInterval) {
            this.lastFuelSpawnY -= this.spawnInterval;

            // 90% chance for at least one
            if (Math.random() < 0.9) {
                const y = this.lastFuelSpawnY - Math.random() * 1000;
                const x = (Math.random() - 0.5) * GameConfig.chunkSize * 3; // Wide spread

                items.push(new FuelItem(x, y));

                // 50% chance for a second one
                if (Math.random() < 0.5) {
                    const y2 = this.lastFuelSpawnY - Math.random() * 1000;
                    const x2 = (Math.random() - 0.5) * GameConfig.chunkSize * 3;
                    items.push(new FuelItem(x2, y2));
                }
            }
        }
        return items;
    }

    generate(shipX: number, shipY: number, existingPlanets: Planet[] = []): Planet[] {
        const newPlanets: Planet[] = [];
        const currentChunkX = Math.floor(shipX / GameConfig.chunkSize);
        const currentChunkY = Math.floor(shipY / GameConfig.chunkSize);
        const minGap = 100;

        // Check 3x3 grid around current chunk
        for (let dx = -1; dx <= 1; dx++) {
            for (let dy = -1; dy <= 1; dy++) {
                const chunkX = currentChunkX + dx;
                const chunkY = currentChunkY + dy;
                const chunkKey = `${chunkX},${chunkY}`;

                if (!this.visitedChunks.has(chunkKey)) {
                    this.visitedChunks.add(chunkKey);

                    // Generate planets in this chunk
                    const count = GameConfig.minPlanetsPerChunk + Math.floor(Math.random() * (GameConfig.maxPlanetsPerChunk - GameConfig.minPlanetsPerChunk));

                    for (let i = 0; i < count; i++) {
                        let attempts = 0;
                        let valid = false;
                        let x = 0;
                        let y = 0;
                        let radius = 0;

                        while (!valid && attempts < 10) {
                            attempts++;
                            // Position within chunk
                            x = (chunkX * GameConfig.chunkSize) + Math.random() * GameConfig.chunkSize;
                            y = (chunkY * GameConfig.chunkSize) + Math.random() * GameConfig.chunkSize;

                            radius = GameConfig.minPlanetRadius + Math.random() * (GameConfig.maxPlanetRadius - GameConfig.minPlanetRadius);

                            // Check collision with ship start (approx)
                            if (chunkX === 0 && chunkY === 0) {
                                if (Math.abs(x) < 350 && Math.abs(y) < 350) continue;
                            }

                            valid = true;

                            // Check collision with newPlanets
                            for (const p of newPlanets) {
                                const dist = Math.sqrt((x - p.x) ** 2 + (y - p.y) ** 2);
                                if (dist < (radius + p.radius + minGap)) {
                                    valid = false;
                                    break;
                                }
                            }
                            if (!valid) continue;

                            // Check collision with existingPlanets
                            for (const p of existingPlanets) {
                                const dist = Math.sqrt((x - p.x) ** 2 + (y - p.y) ** 2);
                                if (dist < (radius + p.radius + minGap)) {
                                    valid = false;
                                    break;
                                }
                            }
                        }

                        if (valid) {
                            // Random Type Logic based on depth
                            let type: PlanetType;
                            const types = Object.values(PlanetType) as PlanetType[];

                            // Filter valid types based on depth (y is negative going up)
                            // 50 LY = -5000
                            // 100 LY = -10000
                            // 200 LY = -20000

                            const depth = -y; // Positive distance

                            // Phase 1: 0-50 LY (No Black Holes)
                            let availableTypes = types.filter(t => t !== PlanetType.Asteroid && t !== PlanetType.Star);
                            if (depth < 5000) {
                                availableTypes = availableTypes.filter(t => t !== PlanetType.BlackHole);
                            }

                            // Phase 3: 200 LY+ (Add Stars)
                            if (depth > 20000) {
                                // Small chance for Star
                                if (Math.random() < 0.05) {
                                     type = PlanetType.Star;
                                     radius = GameConfig.maxPlanetRadius; // Stars are big
                                } else {
                                     type = availableTypes[Math.floor(Math.random() * availableTypes.length)];
                                }
                            } else {
                                type = availableTypes[Math.floor(Math.random() * availableTypes.length)];
                            }

                            newPlanets.push(new Planet(x, y, radius, type));
                        }
                    }

                    // Phase 2: 100 LY+ (Add Asteroid Clusters)
                    // Separate pass for asteroids to allow density
                    const asteroidStartDepth = 10000;
                    const chunkDepth = -chunkY * GameConfig.chunkSize;

                    if (chunkDepth > asteroidStartDepth) {
                         if (Math.random() < 0.4) { // 40% chance per chunk to have a cluster
                             const clusterCount = Math.floor(Math.random() * 5) + 3; // 3-8 asteroids
                             const clusterX = (chunkX * GameConfig.chunkSize) + Math.random() * GameConfig.chunkSize;
                             const clusterY = (chunkY * GameConfig.chunkSize) + Math.random() * GameConfig.chunkSize;

                             for(let k=0; k<clusterCount; k++) {
                                 const offX = (Math.random() - 0.5) * 200;
                                 const offY = (Math.random() - 0.5) * 200;
                                 const ax = clusterX + offX;
                                 const ay = clusterY + offY;
                                 const ar = 10 + Math.random() * 10; // Small radius

                                 // Simple collision check against existing planets only (not other asteroids to allow overlap/density)
                                 let aValid = true;
                                 for (const p of newPlanets) {
                                     const dist = Math.sqrt((ax - p.x)**2 + (ay - p.y)**2);
                                     if (dist < p.radius + ar + 50) {
                                         aValid = false; break;
                                     }
                                 }
                                 if (existingPlanets) {
                                      for (const p of existingPlanets) {
                                         const dist = Math.sqrt((ax - p.x)**2 + (ay - p.y)**2);
                                         if (dist < p.radius + ar + 50) {
                                             aValid = false; break;
                                         }
                                     }
                                 }

                                 if (aValid) {
                                     newPlanets.push(new Planet(ax, ay, ar, PlanetType.Asteroid));
                                 }
                             }
                         }
                    }
                }
            }
        }
        return newPlanets;
    }

    // Remove visited chunks that are far away to allow regeneration if returned to
    cleanup(shipX: number, shipY: number, cleanupRadius: number) {
        const currentChunkX = Math.floor(shipX / GameConfig.chunkSize);
        const currentChunkY = Math.floor(shipY / GameConfig.chunkSize);
        const cleanupChunkRadius = Math.ceil(cleanupRadius / GameConfig.chunkSize);

        for (const key of this.visitedChunks) {
            const [cxStr, cyStr] = key.split(',');
            const cx = parseInt(cxStr);
            const cy = parseInt(cyStr);

            // Simple Manhattan or Chebyshev distance check is faster
            if (Math.abs(cx - currentChunkX) > cleanupChunkRadius ||
                Math.abs(cy - currentChunkY) > cleanupChunkRadius) {
                this.visitedChunks.delete(key);
            }
        }
    }

    reset(startY: number = 0) {
        this.visitedChunks.clear();
        this.lastFuelSpawnY = startY;
    }
}

File: python_project/level_generator.py
import math
import random
from game_config import GameConfig
from entities import Planet, PlanetType, FuelItem

class LevelGenerator:
    def __init__(self):
        self.visitedChunks = set()
        self.lastFuelSpawnY = 0
        self.spawnInterval = GameConfig.fuelSpawnInterval

    def generateItems(self, shipY):
        items = []
        if shipY < self.lastFuelSpawnY - self.spawnInterval:
            self.lastFuelSpawnY -= self.spawnInterval

            if random.random() < 0.9:
                y = self.lastFuelSpawnY - random.random() * 1000
                x = (random.random() - 0.5) * GameConfig.chunkSize * 3
                items.append(FuelItem(x, y))

                if random.random() < 0.5:
                    y2 = self.lastFuelSpawnY - random.random() * 1000
                    x2 = (random.random() - 0.5) * GameConfig.chunkSize * 3
                    items.append(FuelItem(x2, y2))
        return items

    def generate(self, shipX, shipY, existingPlanets=None):
        if existingPlanets is None:
            existingPlanets = []

        newPlanets = []
        currentChunkX = math.floor(shipX / GameConfig.chunkSize)
        currentChunkY = math.floor(shipY / GameConfig.chunkSize)
        minGap = 100

        for dx in range(-1, 2):
            for dy in range(-1, 2):
                chunkX = currentChunkX + dx
                chunkY = currentChunkY + dy
                chunkKey = f"{chunkX},{chunkY}"

                if chunkKey not in self.visitedChunks:
                    self.visitedChunks.add(chunkKey)

                    count = GameConfig.minPlanetsPerChunk + math.floor(random.random() * (GameConfig.maxPlanetsPerChunk - GameConfig.minPlanetsPerChunk))

                    for _ in range(count):
                        attempts = 0
                        valid = False
                        x = 0
                        y = 0
                        radius = 0

                        while not valid and attempts < 10:
                            attempts += 1
                            x = (chunkX * GameConfig.chunkSize) + random.random() * GameConfig.chunkSize
                            y = (chunkY * GameConfig.chunkSize) + random.random() * GameConfig.chunkSize
                            radius = GameConfig.minPlanetRadius + random.random() * (GameConfig.maxPlanetRadius - GameConfig.minPlanetRadius)

                            if chunkX == 0 and chunkY == 0:
                                if abs(x) < 350 and abs(y) < 350:
                                    continue

                            valid = True

                            for p in newPlanets:
                                dist = math.sqrt((x - p.x)**2 + (y - p.y)**2)
                                if dist < (radius + p.radius + minGap):
                                    valid = False
                                    break

                            if not valid:
                                continue

                            for p in existingPlanets:
                                dist = math.sqrt((x - p.x)**2 + (y - p.y)**2)
                                if dist < (radius + p.radius + minGap):
                                    valid = False
                                    break

                        if valid:
                            depth = -y
                            availableTypes = [PlanetType.Dead, PlanetType.Ocean, PlanetType.Ice, PlanetType.Desert, PlanetType.GasGiant, PlanetType.BlackHole]

                            if depth < 5000:
                                availableTypes.remove(PlanetType.BlackHole)

                            if depth > 20000:
                                if random.random() < 0.05:
                                    ptype = PlanetType.Star
                                    radius = GameConfig.maxPlanetRadius
                                else:
                                    ptype = random.choice(availableTypes)
                            else:
                                ptype = random.choice(availableTypes)

                            newPlanets.append(Planet(x, y, radius, ptype))

                    asteroidStartDepth = 10000
                    chunkDepth = -chunkY * GameConfig.chunkSize

                    if chunkDepth > asteroidStartDepth:
                        if random.random() < 0.4:
                            clusterCount = math.floor(random.random() * 5) + 3
                            clusterX = (chunkX * GameConfig.chunkSize) + random.random() * GameConfig.chunkSize
                            clusterY = (chunkY * GameConfig.chunkSize) + random.random() * GameConfig.chunkSize

                            for _ in range(clusterCount):
                                offX = (random.random() - 0.5) * 200
                                offY = (random.random() - 0.5) * 200
                                ax = clusterX + offX
                                ay = clusterY + offY
                                ar = 10 + random.random() * 10

                                aValid = True
                                for p in newPlanets:
                                    dist = math.sqrt((ax - p.x)**2 + (ay - p.y)**2)
                                    if dist < p.radius + ar + 50:
                                        aValid = False
                                        break

                                if aValid and existingPlanets:
                                    for p in existingPlanets:
                                        dist = math.sqrt((ax - p.x)**2 + (ay - p.y)**2)
                                        if dist < p.radius + ar + 50:
                                            aValid = False
                                            break

                                if aValid:
                                    newPlanets.append(Planet(ax, ay, ar, PlanetType.Asteroid))
        return newPlanets

    def cleanup(self, shipX, shipY, cleanupRadius):
        currentChunkX = math.floor(shipX / GameConfig.chunkSize)
        currentChunkY = math.floor(shipY / GameConfig.chunkSize)
        cleanupChunkRadius = math.ceil(cleanupRadius / GameConfig.chunkSize)

        keysToRemove = []
        for key in self.visitedChunks:
            cxStr, cyStr = key.split(',')
            cx = int(cxStr)
            cy = int(cyStr)

            if abs(cx - currentChunkX) > cleanupChunkRadius or abs(cy - currentChunkY) > cleanupChunkRadius:
                keysToRemove.append(key)

        for key in keysToRemove:
            self.visitedChunks.remove(key)

    def reset(self, startY=0):
        self.visitedChunks.clear()
        self.lastFuelSpawnY = startY
